[#65875] Collaborative real time editing in documents

ba1ash opened the pull request:
https://community.openproject.org/work_packages/65875
https://community.openproject.org/wp/66328

ba1ash committed:
[#65875] Enable collaborative editing.

ba1ash committed:
[#65875] Enable comments.

ba1ash committed:
Adjust collaborative editing.

- Fix initial content loading.
- Enable commenting on documents.
- Enable persistence on hocuspocus server

ba1ash committed:
Merge remote-tracking branch 'origin/dev' into feature/65875-collaborative-real-time-editing-in-documents

akabiru commented on app/services/collaborative_editing.rb:
```diff
@@ -0,0 +1,57 @@
+          {
+            document_id:,
+            document_text:,
+            exp: 20.minutes.from_now.to_i
```
🟠 Q: For my understanding, is there a specific reason for setting 20 minutes as the expiration period or is it an arbitrary projection?

ba1ash replied:
No, I just picked a more or less reasonable time. It is important to have it set. Can be adjusted.
It is a certain level of protection for a token leakage. So, it is not possible to use it forever to get access to the document.

akabiru commented on docker/dev/hocuspocus/app/index.js:
```diff
@@ -0,0 +1,68 @@
+import { Server } from "@hocuspocus/server";
+import { createVerifier } from 'fast-jwt'
+import { ServerBlockNoteEditor } from "@blocknote/server-util";
+import {
+  BlockNoteSchema,
+  defaultBlockSpecs,
+  defaultInlineContentSpecs,
+  defaultStyleSpecs,
+} from "@blocknote/core";
```
Worth addressing this?

akabiru commented on docker/dev/hocuspocus/app/index.js:
```diff
@@ -0,0 +1,68 @@
+  defaultInlineContentSpecs,
+  defaultStyleSpecs,
+} from "@blocknote/core";
+import * as Y from "yjs";
```
ditto:

akabiru commented on docker/dev/hocuspocus/app/index.js:
```diff
@@ -0,0 +1,68 @@
+  },
+});
+
+server.listen();
```
Q: I understand this is to run a local dev instance- it's unclear to me how/where the production instance will be configured, and whether the logic in the hooks here will be applied.

I'm also missing out on some level of documentation on the general flow of things. I think it would go a long way to help everyone understand the setup.

ba1ash replied:
I removed this code completely in favor of image published from https://github.com/opf/op-blocknote-hocuspocus.
I will add some documentation to this repo.

akabiru commented on frontend/src/global_styles/openproject.sass:
```diff
@@ -29,3 +29,22 @@
 // Component specific Styles
 @import "../../../app/components/_index.sass"
 @import "../../../app/forms/_index.sass"
+
+.comments-main-container
+  align-items: center
+  display: flex
+  flex-direction: column-reverse
+  gap: 10px
+  height: 100%
+  max-width: none
+  padding: 10px
+  width: 100%
+  .bn-editor
+    height: 100%
+    max-width: 700px
+    min-height: 80vh
+    overflow: auto
+    width: 100%
+  .bn-thread-comments
+    .bn-editor
+      min-height: 3vh
```
🟠 Shall we put these overrides in a dedicated file? Contextualize it by the file name and add some top level comments as to why the override is necessary.

ba1ash replied:
This one has been removed. No need for it, actually.

akabiru commented on frontend/src/react/OpBlockNoteContainer.tsx:
```diff
@@ -50,10 +62,68 @@ const schema = BlockNoteSchema.create({
+    const cursorColor = '#' + Math.floor(Math.random() * 16777215).toString(16).padStart(6, '0');
+    collaboration = {
+      provider: hocuspocusProvider,
+      fragment: doc.getXmlFragment("document-store"),
+      user: {
+        name: activeUser.username,
+        color: cursorColor,
+      },
+      showCursorLabels: "activity"
+    }
+    threadStore = new YjsThreadStore(
+      activeUser.id,
+      doc.getMap("threads"),
+      new DefaultThreadStoreAuth(activeUser.id, "editor"),
+    );
+    comments = {
+      threadStore: threadStore,
+    };
+  }
+
+  let editor: any;
+  if(collaborationEnabled) {
+    const resolveUsers = async (userIds: string[]) => {
+      return users.filter((user) => userIds.includes(user.id));
+    }
+
+    editor = useCreateBlockNote(
+      {
+        resolveUsers,
+        collaboration,
+        schema,
+        comments
+      },
+      [activeUser, threadStore]
+    );
+  } else {
+    editor = useCreateBlockNote(
+      { schema },
+    );
+  };
```
🟠 Perhaps soon we can look into modularizing this component; the sheer mass of it is overwhelming. cc @brunopagno

ba1ash replied:
Yeah, it makes sense...

akabiru reviewed (commented):
Nicely done, Pavel 👍🏾 . I left some remarks for your consideration. In general, I'm missing out on some level of documentation. I think it's important to document sooner rather than later.

ba1ash committed:
Merge remote-tracking branch 'origin/dev' into feature/65875-collaborative-real-time-editing-in-documents

ba1ash committed:
[#65875] Refactor collaborative editing.

- Use published docker image for local dev setup.
- Handle a case when Setting.collaborative_editing_hocuspocus_url is set to invalid URI.
- Remove unneedd CSS.
- Add some tests.

akabiru commented on app/controllers/concerns/dynamic_content_security_policy.rb:
```diff
@@ -47,4 +53,23 @@ def append_content_security_policy_directives(directives)
+    if hocuspocus_url.present?
+      uri = begin
+        URI.parse(hocuspocus_url)
+      rescue URI::InvalidURIError
+        OpenProject.logger.info do
+          "Setting.collaborative_editing_hocuspocus_url is set to an invalid URI: #{hocuspocus_url}"
+        end
```
Nice 👍🏾

akabiru commented on frontend/package.json:
```diff
@@ -60,9 +60,10 @@
     "theo": "^8.1.5",
     "ts-node": "~8.3.0",
-    "typescript": "^5.8.3",
+    "typescript": "5.8.3",
```
```suggestion
    "typescript": "^5.8.3",
```

🟠 Was it intentional to lock in typescript to v5.8.3?

ba1ash replied:
No, it was not intentional.

akabiru reviewed (approved):
Structure looks good, thanks for adding tests! Documentation remains an important topic from my POV, please track that somewhere. I would also be pleased if we addressed my earlier questions marked in 🟠

ba1ash committed:
Remove exact version locking for typescript in package.json.

Co-authored-by: Kabiru Mwenja <k.mwenja@openproject.com>

myabc commented on frontend/package.json:
```diff
@@ -60,7 +60,7 @@
     "theo": "^8.1.5",
     "ts-node": "~8.3.0",
-    "typescript": "5.8.3",
+    "typescript": "^5.8.3",
```
Please remember to commit the `package-lock.json` as well! 1cfcb768
